File: 227. Basic Calculator II.py
class Solution:
    def calculate(self, s: str) -> int:
        stack = []
        idx = 0
        sign = 1
        output = 0
        number = 0
        while idx < len(s):
            char = s[idx]
            if char == ' ':
                idx += 1
            elif char == '+':
                output += sign * number
                number = 0
                sign = 1
                idx += 1
            elif char == '-':
                output += sign * number
                number = 0
                sign = -1
                idx += 1
            elif char == '*':
                temp = 0
                idx += 1
                while idx < len(s) and (s[idx] == ' ' or s[idx].isdigit()):
                    if s[idx] == ' ':
                        idx += 1
                    else:
                        temp = temp*10 + int(s[idx])
                        idx += 1
                number *= temp
            elif char == '/':
                temp = 0
                idx += 1
                while idx < len(s) and (s[idx] == ' ' or s[idx].isdigit()):
                    if s[idx] == ' ':
                        idx += 1
                    else:
                        temp = temp*10 + int(s[idx])
                        idx += 1
                if number * temp < 0:
                    number = number // temp + 1
                else:
                    number = number // temp
            elif char.isdigit():
                number = number*10 + int(char)
                idx += 1
                while idx< len(s) and s[idx].isdigit():
                    number = number*10 + int(s[idx])
                    idx += 1
        return output + sign * number

File: 227. Basic Calculator II.java
/**
 * A more compact version
 * Also O(1) Space
 */
class Solution {
    public int calculate(String s) {
        int result=0, curr=0, temp=0;   
        char lastSign = '+';
        for (int i=0; i<s.length(); i++) {
            char c = s.charAt(i);
            if (Character.isDigit(c)) {
                curr = curr * 10 + (c - '0');
            } else if (c != ' ') {
                switch (lastSign) {
                    case '+':
                        result += temp;
                        temp = curr;
                        break;
                    case '-':
                        result += temp;
                        temp = -curr;
                        break;
                    case '*':
                        temp *= curr;
                        break;
                    case '/':
                        temp /= curr;
                        break;
                }
                lastSign = c;
                curr = 0;
            }
        }
        switch (lastSign) {
            case '+':
                result += temp;
                temp = curr;
                break;
            case '-':
                result += temp;
                temp = -curr;
                break;
            case '*':
                temp *= curr;
                break;
            case '/':
                temp /= curr;
                break;
        }
        result += temp;
        return result;
    }
}
/**
 * My solution without stack
 * Time O(n) Space O(1)
 */
class Solution {
    public int calculate(String s) {
        int result=0, curr=0, temp=0;        
        boolean isPlus = true;
        boolean isTimes = true;
        boolean containsMulti = false;
        boolean multiplyWithZero = false;
        int i=0;
        while (i < s.length()) {
            char c = s.charAt(i);
            if (Character.isDigit(c)) {
                curr = curr * 10 + (c-'0');
            } else if (c == '+' || c == '-') {
                if (!multiplyWithZero) {
                    if (containsMulti) {
                        if (isTimes) {
                            curr = temp * curr;
                        } else {
                            curr = temp / curr;
                        }
                    }
                    if (isPlus) {
                        result += curr;
                    } else {
                        result -= curr;
                    }
                }
                multiplyWithZero = false;
                containsMulti = false;
                isPlus = (c == '+');
                curr = 0;
                temp = 0;
            } else if (c == '*' || c == '/') {
                if (multiplyWithZero) {
                    temp = 0;
                } else {
                    if (curr == 0) {
                        multiplyWithZero = true;
                    }
                    if (containsMulti) {
                        if (isTimes) {
                            temp *= curr;
                        } else {
                            temp /= curr;
                        }
                    } else {
                        temp = curr;
                    }
                    
                }
                curr = 0;
                isTimes = (c == '*');
                containsMulti = true;
            }
            i++;
        }
        if (!multiplyWithZero) {
            if (containsMulti) {
                if (isTimes) {
                    curr = temp * curr;
                } else {
                    curr = temp / curr;
                }
            }
            if (isPlus) {
                result += curr;
            } else {
                result -= curr;
            }
        }
        return result;
    }
}